client/shared/src/api/util.ts

```
@@ -64,6 +64,39 @@ export const isSubscribable = (value: unknown): value is Subscribable<unknown> =
     hasProperty('subscribe')(value) &&
     typeof value.subscribe === 'function'
 
+/**
+ * Reports whether the value is an AsyncIterable
+ */
+export const isAsyncIterable = (value: unknown): value is AsyncIterable<unknown> =>
+    typeof value === 'object' && value !== null && typeof (value as any)[Symbol.asyncIterator] === 'function'
+
+/**
+ * Convert an async iterable into an observable.
+ *
+ * @param iterable The source iterable.
+ */
+export const observableFromAsyncIterable = <T>(iterable: AsyncIterable<T>): Observable<T> =>
+    new Observable((observer: Observer<T>) => {
+        let unsubscribed = false
+
+        ;(async () => {
+            for await (const value of iterable) {
+                if (unsubscribed) {
+                    // Swallow latest value if unsubscribed, stop iteration
+                    return
+                }
+
+                observer.next(value)
+            }
+        })()
+            .catch(error => observer.error(error))
+            .finally(() => observer.complete())
```

Note: Only one of `complete` or `error` can be called, both will terminate the stream. I.e. `complete` is not called after `error`.

---

Is there any harm in calling `complete` anyways? For clarity, I'll move if after the for await...of loop

---

Probably not if RxJS (or whatever Subscribable we receive) is coded defensively enough to silently drop the call (and not error, nor pass it through). But it could create a false impression when reading this code.